src/cli_tool/caw_cli.go

```diff
@@ -0,0 +1,96 @@
+	userPtr         = flag.String("user", "", "Logs in as the given username")
+	cawPtr          = flag.String("caw", "", "Creates a new Caw with the given text")
+	replyPtr        = flag.Int("reply", -1,
+														 "Indicates that the new caw is a reply to the given id")
+	followPtr  = flag.String("follow", "", "Starts following the given username")
+	readPtr    = flag.Int("read", -1, "Reads the caw thread starting at the given id")
+	profilePtr = flag.Bool("profile", false,
+												 "Gets the user’s profile of following and followers")
+	hookPtr      = flag.String("hook", "", "Hooks the given function")
+	unhookPtr    = flag.String("unhook", "", "Unhooks the given function")
+	hookallPtr   = flag.Bool("hookall", false, "Hooks all stored functions")
+	unhookallPtr = flag.Bool("unhookall", false, "Unhooks all stored functions")
+)
+
+func main() {
+	// Checking and handling arguments
+	if len(os.Args) < 2 {
+		fmt.Println("You have not used any command line arguments. Please try again.")
+		flag.PrintDefaults()
+		os.Exit(1)
+	}
+
+	flag.Parse()
+
+	// Ensuring a username is always specified
+	// With exception of registeruser
+	if *registeruserPtr == "" && *userPtr == "" && *hookPtr == "" &&
+		*unhookPtr == "" && *unhookallPtr == false &&
+		*hookallPtr == false {
+		err := fmt.Errorf("You cannot perform any action other than " +
+											"hooking/unhooking without specifying a username. \n" +
+											"Please consider using the --user <username> flag.")
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+
+	if *hookallPtr && *unhookallPtr {
+		err := fmt.Errorf("You cannot hook all and unhook all simultaneously.")
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+
+	if *replyPtr != -1 && *cawPtr == "" {
+		err := fmt.Errorf("You cannot reply without specifying a caw. \n" +
+											"Please consider using the --caw <caw text> flag.")
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+
+	// Error checking done, now
+	// we can handle arguments
+
+	// Creating client instance
+	client := caw.CreateCawClient()
+
+	if *registeruserPtr != "" {
+		client.RegisterUser(*registeruserPtr)
+	} else if *replyPtr != -1 {
+		client.PostCaw(*userPtr, *cawPtr, *replyPtr)
+	} else if *cawPtr != "" {
+		// Since we have already checked for reply
+		// We know this would be a caw and not a reply
+		// According to specs, we specify a new caw using -1
+		client.PostCaw(*userPtr, *cawPtr, -1)
+	} else if *followPtr != "" {
+		client.FollowUser(*userPtr, *followPtr)
+	} else if *readPtr != -1 {
+		client.ReadCaw(*readPtr)
+	} else if *profilePtr {
+		client.GetProfile(*userPtr)
+	} else if *hookPtr != "" {
+		client.HookFunction(*hookPtr)
+	} else if *unhookPtr != "" {
+		client.UnhookFunction(*unhookPtr)
+	} else if *hookallPtr {
+		client.HookAll()
+	} else if *unhookallPtr {
+		client.UnhookAll()
+	}
+	
+	// Calling destructor to close Faz connection
+	client.DestroyCawClient()
```

Comment:
Do you know if this destroy idea is common in Go?

Reply:
Destructors are a debated topic in Go as far as I have read. What some people do is to declare a similar method called Close() and as soon as the constructor method is called, a following "defer "type".Close()" method is called. In general people are pushing for more formal destructors for better resource management etc., but for now hacks like these will have to do. 

You can read up on this more here if you're interested! https://github.com/golang/go/issues/38057